pyanaconda/ui/webui/__init__.py

```diff
@@ -75,11 +74,6 @@ def setup(self, data):
 
         This method must be provided by all subclasses.
         """
-        # Finish all initialization jobs.
-        # FIXME: Control the initialization via DBus.
-        self._print_message("Waiting for all threads to finish...")
-        thread_manager.wait_all()
```

Comment:
Are we sure this is really not necessary, eq. not to get for example storage (and in the future RPM installation sources) read i to the Web UI in some undefined stage ?

Reply:
You don't need that because you don't executing any code here. The only thing which happens after this in the main thread is wait. The logic is already moved to modules or JS in this case.

I can't proof it but I don't see any reason to keep it here.

Reply:
This makes sure that Anaconda modules have finished all started tasks and are fully initialized. While GUI is able to start GTK and handle the initialization threads in spokes, I don't think that Web UI has any functionality that would handle that.

Reply:
@poncovka the way we handle this in the web UI, is that we check the /run/anaconda/bus-address file, and when this get's written we initialize all dbus APIs clients, add event listeners (for some properties we look at) and fetch some initialization data for the UI. If some of these returns an error we show and 'Critical error' dialog.

I just noticed that if the client disconnects we just show an error in the console but we don't quit the installation. I guess we should quite the installation with the usual 'Critical error' dialog. https://github.com/rhinstaller/anaconda/blob/master/ui/webui/src/apis/storage.js#L45 

But if I understand correctly what this code does, I think this is handled in the UI as expected.

Any comments?

Reply:
@KKoukiou That will handle errors of DBus calls made by the Web UI. I am talking about handling errors of DBus calls made by the `anaconda.py` script and all threads started by this script. This code handles initialization of the modules based on the boot options, cmdline options, configuration files and the `interactive-defaults.ks` kickstart file. It is important to make sure that this is finished before the user can interact with the UI and start to modify the state of the modules. Otherwise, the user and the threads will do these modifications at the same time.

In TUI and GUI, the synchronization of the initialization steps is handled in the spokes via the thread management. The user is not allowed to enter the spoke if there is still a running initialization thread related to this spoke. In Web UI, this is not an option. I would recommend the following approach.

1. Always make sure that all threads are finished before you start the Web UI. This is the current implementation and the easiest solution. Yes, it is blocking the installation and it can take longer to start the Web UI, but it won't affect automated installations where the speed could really matter plus I have never seen a requirement to optimize the speed of the installation. Personally, I would invest time into adding a splash screen and eventually figuring out how to show some kind of a progress reporting instead of re-introducing the parallelism. In Anaconda, it is a common source of difficult bugs.

2. Optimize intentionally. Instead of ignoring all initialization threads, let's investigate which threads are the most blocking and figure out how to bypass these. The script can skip some of the initializations steps if Web UI will take responsibility for handling these. I guess that the first storage rescan might be the most problematic, so maybe it would make sense to focus on that one.

3. Modularize the initialization. The content of the `anaconda.py` script could be eventually moved on DBus to the Runtime module or the Boss. The idea is to reuse the installation queue, move the initialization steps to tasks (if they are not tasks already) and run the initialization in a similar way as the installation. This was always a nice to have thing and we never did any scoping of this, so I don't think it worth it to rely on this solution at this moment. It will take time.

Reply:
> I think I thought we are already on level that web UI is checking modules initialization.

Your understanding and expectations of the current state of the modularization scare me. I tried to be very clear that that the runtime configuration (that includes the initialization) is not modularized. There are issues and epics for that. It is not finished, because it required all other modules to be fully migrated. The last work I did was to complete the migration of the Payloads module. That's it.

At this moment, the D-Bus modules provide enough support for the user interfaces to collect the user preferences and install a system based on them. That's all. The initialization code it not modularized. It eventually uses the DBus API of modules to set them up, but the whole logic for that is implemented and driven by the `anaconda.py` script.

The modularization of the initialization code won't be easy. First of all, it will be necessary to create new initialization tasks for all the steps that are happening in the `anaconda.py` script. Then you can schedule and run these tasks in a similar way as the installation tasks via a queue. At this point, Web UI will be able to monitor the initialization. Unfortunately, this also means that the initialization threads are no longer usable in TUI and GUI, so it will be necessary to completely rewrite conditions for the initialization and readiness of every spoke. It will take months.

> However, on second thought, I don't think that return of the `wait_all()` call will help. The issue is that we already started FF in the `liveinst` script, so this will just postpone when `anaconda.py` starts to monitor it not really when Firefox or UI is started. From that purpose, it seems that this PR is wrong completely and I don't see a way how to improve the current state much. We can try for (2) but I don't have big hopes for that. Also (2) seems to me like a fragile solution because it will be hard to find out what could cause troubles if skipped.

I think you should add some kind of a splash screen and start FF with that one. This screen should be visible until the back-end is ready for any user interaction. The problem isn't that the initialization takes time. The problem is that it is not properly communicated to users.

> More and more I'm thinking about this then I think the web UI needs to keep an eye about the initialization not `anaconda.py` or maybe both of them?

How?

Reply:
We discussed this solution and it seems the best approach will be to introduce a flag file solution as signal to web UI that the backend is ready and can be used. This flag file will be created after the waiting is finished.